### PR TITLE
[ES|QL] Add CombineBinaryComparisons rule

### LINKED ISSUE #108525: [CI] LogicalPlanOptimizerTests testSimplifyComparisonArithmeticWithDisjunction failing
I haven't investigated this one yet.  There are several other issues around how this optimization works, and it's possible fixing one of them will fix this test.

**Build scan:**
https://gradle-enterprise.elastic.co/s/zn7kv6lwmdrow/tests/:x-pack:plugin:esql:test/org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests/testSimplifyComparisonArithmeticWithDisjunction

**Reproduction line:**
```
./gradlew ':x-pack:plugin:esql:test' --tests "org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.testSimplifyComparisonArithmeticWithDisjunction" -Dtests.seed=A8AF70E4F3429963 -Dtests.locale=zh-TW -Dtests.timezone=Asia/Chungking -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Yes

**Failure history:**
[Failure dashboard for `org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests#testSimplifyComparisonArithmeticWithDisjunction`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testSimplifyComparisonArithmeticWithDisjunction'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure excerpt:**
```
java.lang.AssertionError: Expected filter condition to be a binary comparison but found [12 * (-integer - 5) >= -120 OR integer < 5]

  at __randomizedtesting.SeedInfo.seed([A8AF70E4F3429963:7A91B4839C45DBAE]:0)
  at org.junit.Assert.fail(Assert.java:89)
  at org.junit.Assert.assertTrue(Assert.java:42)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.extractPlannedBinaryComparison(LogicalPlanOptimizerTests.java:4405)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.extractPlannedBinaryComparison(LogicalPlanOptimizerTests.java:4397)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.doTestSimplifyComparisonArithmetics(LogicalPlanOptimizerTests.java:4418)
  at org.elasticsearch.xpack.esql.optimizer.LogicalPlanOptimizerTests.testSimplifyComparisonArithmeticWithDisjunction(LogicalPlanOptimizerTests.java:4512)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at org.junit.rules.RunRules.evaluate(RunRules.java:20)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

### PR DESCRIPTION
**Pull request by fang-xing-esql**
Resolves #108525

This rule exists in sql, but not esql. After adding this rule the OR predicate in the test can be transformed into a boolean term. The sql rule is modified as ES|QL doesn't support range yet.

### CODE REVIEW

**Review by astefan — CHANGES_REQUESTED** [review]
I don't believe this change is enough.
This rule you are adding (from SQL) had tests in `OptimizerRulesTests`. We must add as many as possible back in ESQL.

**Comment by astefan** [comment]
Thinking about this some more, I _think_ we didn't add this rule because of multi-values fields. EQL doesn't use this rule and EQL is multi-values aware. SQL, on the other hand, is single-value enforcer and uses it.

Please, look into the multi-values issue and see if this rule makes sense in this scenario.

**Comment by fang-xing-esql** [comment]
> Thinking about this some more, I _think_ we didn't add this rule because of multi-values fields. EQL doesn't use this rule and EQL is multi-values aware. SQL, on the other hand, is single-value enforcer and uses it.
> 
> Please, look into the multi-values issue and see if this rule makes sense in this scenario.

I'll look into multi-valued fields and migrating more tests, there must be a reason that this rule was not added for ES|QL in the first place.

**Comment by fang-xing-esql** [comment]
> I don't believe this change is enough. This rule you are adding (from SQL) had tests in `OptimizerRulesTests`. We must add as many as possible back in ESQL.

Thanks for reviewing @astefan ! More tests(from `OptimizerRulesTests`) related to the `CombineBinaryComparisons` rule are added into esql in `CombineBinaryComparisonsTests`. The tests related to `Range` are not added, as Range is not supported in esql yet. The BinaryComparison operators the rule affects are ==, !=, >, >=, <, <=, they work with single value fields, they throw warnings if the fields are multi-valued. I did some tests on multi-valued fields, they seem work as expected.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ints.csv-spec
Comment (astefan) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/ints.csv-spec:1007]: Please, add this test as well `row x = [1,2,3] | where 12 * (-x - 5) >= -120 OR x < 5`. It's the same formula that this PR tries to fix.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java
Comment (astefan) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:4617]: Apart from this test, I want to see another test that actually shows the effects of adding this rule in the optimizer.
For example, I tested this query `from employees | where 12 * (-languages - 5) >= -120 OR languages < 5 | keep languages` and I looked at how the rules are applied.

Before this rule was added the simplified change to the plan was like this:

```
Rule rules.SimplifyComparisonsArithmetics applied
Limit[1000[INTEGER]]                                                                                           = Limit[1000[INTEGER]]
\_EsqlProject[[languages{f}#172]]                                                                              = \_EsqlProject[[languages{f}#172]]
  \_Filter[NEG(languages{f}#172) - 5[INTEGER] * 12[INTEGER] >= -120[INTEGER] OR languages{f}#172 < 5[INTEGER]] !   \_Filter[NEG(languages{f}#172) - 5[INTEGER] >= -10[INTEGER] OR languages{f}#172 < 5[INTEGER]]
    \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]                             =     \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]

Rule rules.SimplifyComparisonsArithmetics applied
EsqlProject[[languages{f}#172]]                                                                 = EsqlProject[[languages{f}#172]]
\_Limit[1000[INTEGER]]                                                                          = \_Limit[1000[INTEGER]]
  \_Filter[NEG(languages{f}#172) - 5[INTEGER] >= -10[INTEGER] OR languages{f}#172 < 5[INTEGER]] !   \_Filter[NEG(languages{f}#172) >= -5[INTEGER] OR languages{f}#172 < 5[INTEGER]]
    \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]              =     \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]

Rule rules.SimplifyComparisonsArithmetics applied
EsqlProject[[languages{f}#172]]                                                    = EsqlProject[[languages{f}#172]]
\_Limit[1000[INTEGER]]                                                             = \_Limit[1000[INTEGER]]
  \_Filter[NEG(languages{f}#172) >= -5[INTEGER] OR languages{f}#172 < 5[INTEGER]]  !   \_Filter[languages{f}#172 <= 5[INTEGER] OR languages{f}#172 < 5[INTEGER]]
    \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...] =     \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]
```

And, in the end, before all transformations are done, I am seeing the final query that gets pushed to ES:

```
ProjectExec[[languages{f}#17]]                                                       = ProjectExec[[languages{f}#17]]
\_LimitExec[1000[INTEGER]]                                                           ! \_FieldExtractExec[languages{f}#17]<[]>
  \_FilterExec[languages{f}#17 <= 5[INTEGER] OR languages{f}#17 < 5[INTEGER]]        !   \_EsQueryExec[employees], indexMode[standard], query[{"bool":{"should":[{"esql_single_value":{"field":"languages","next":{"range":{"languages":{"lte":5,"boost":1.0}}},"source":"12 * (-languages - 5) >= -120@1:24"}},{"esql_single_value":{"field":"languages","next":{"range":{"languages":{"lt":5,"boost":1.0}}},"source":"languages < 5@1:57"}}],"boost":1.0}}][_doc{f}#21], limit[1000], sort[] estimatedRowSize[null]
    \_EsSourceExec[employees][languages{f}#17, languages.byte{f}#18, languages.lo..] !
```

And you see there two `range` queries (in `EsQueryExec`).

After this rule is being added to the optimizer, the simplified chain of rules is like this:


```
Rule rules.SimplifyComparisonsArithmetics applied
Limit[1000[INTEGER]]                                                                                           = Limit[1000[INTEGER]]
\_EsqlProject[[languages{f}#172]]                                                                              = \_EsqlProject[[languages{f}#172]]
  \_Filter[NEG(languages{f}#172) - 5[INTEGER] * 12[INTEGER] >= -120[INTEGER] OR languages{f}#172 < 5[INTEGER]] !   \_Filter[NEG(languages{f}#172) - 5[INTEGER] >= -10[INTEGER] OR languages{f}#172 < 5[INTEGER]]
    \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]                             =     \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]

Rule rules.SimplifyComparisonsArithmetics applied
EsqlProject[[languages{f}#172]]                                                                 = EsqlProject[[languages{f}#172]]
\_Limit[1000[INTEGER]]                                                                          = \_Limit[1000[INTEGER]]
  \_Filter[NEG(languages{f}#172) - 5[INTEGER] >= -10[INTEGER] OR languages{f}#172 < 5[INTEGER]] !   \_Filter[NEG(languages{f}#172) >= -5[INTEGER] OR languages{f}#172 < 5[INTEGER]]
    \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]              =     \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]

Rule rules.SimplifyComparisonsArithmetics applied
EsqlProject[[languages{f}#172]]                                                    = EsqlProject[[languages{f}#172]]
\_Limit[1000[INTEGER]]                                                             = \_Limit[1000[INTEGER]]
  \_Filter[NEG(languages{f}#172) >= -5[INTEGER] OR languages{f}#172 < 5[INTEGER]]  !   \_Filter[languages{f}#172 <= 5[INTEGER] OR languages{f}#172 < 5[INTEGER]]
    \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...] =     \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]

Rule rules.CombineBinaryComparisons applied
EsqlProject[[languages{f}#172]]                                                    = EsqlProject[[languages{f}#172]]
\_Limit[1000[INTEGER]]                                                             = \_Limit[1000[INTEGER]]
  \_Filter[languages{f}#172 <= 5[INTEGER] OR languages{f}#172 < 5[INTEGER]]        !   \_Filter[languages{f}#172 <= 5[INTEGER]]
    \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...] =     \_EsRelation[employees][languages{f}#172, languages.byte{f}#173, languages...]
```

So, it is clearly visible in the `EsRelation` that the addition of CombineBinaryComparisons made the actual query that gets pushed to ES be a simpler one (`_Filter[languages{f}#172 <= 5[INTEGER]` instead of `_Filter[languages{f}#172 <= 5[INTEGER] OR languages{f}#172 < 5[INTEGER]`).
I think such a test (that I am assuming we don't have, sorry I haven't checked) would be valuable.

**Review by astefan — COMMENTED** [review]
It seems to be ok with multi-value fields, indeed. Thank you for checking.
I've left two more comments related to a better testing coverage.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:4617]: Yeah, this is exactly what this rule does, after it is applied, there is only one predicate(`EsqlBinaryComparison`) in the Filter in stead of two(`EsqlBinaryComparison` connected by Or) on languages. 

And the test verifies it as you want it to be :) `doTestSimplifyComparisonArithmetics` calls `extractPlannedBinaryComparison` to verify the top level filter has a `EsqlBinaryComparison` directly under it. 

```
    private void doTestSimplifyComparisonArithmetics(
        String expression,
        String fieldName,
        EsqlBinaryComparison.BinaryComparisonOperation opType,
        Object bound
    ) {
        EsqlBinaryComparison bc = extractPlannedBinaryComparison(expression);
        ...
    }
```
```
    private static EsqlBinaryComparison extractPlannedBinaryComparison(LogicalPlan plan) {
        assertTrue("Expected unary plan, found [" + plan + "]", plan instanceof UnaryPlan);
        UnaryPlan unaryPlan = (UnaryPlan) plan;
        assertTrue("Epxected top level Filter, foung [" + unaryPlan.child().toString() + "]", unaryPlan.child() instanceof Filter);
        Filter filter = (Filter) unaryPlan.child();
        assertTrue(
            "Expected filter condition to be a binary comparison but found [" + filter.condition() + "]",
            filter.condition() instanceof EsqlBinaryComparison
        );
        return (EsqlBinaryComparison) filter.condition();
    }
```

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/ints.csv-spec
Comment (fang-xing-esql) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/ints.csv-spec:1007]: Will add it.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java
Comment (astefan) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java:4617]: Yep, you are right. The test I had in mind does basically the same thing.

**Review by astefan — APPROVED** [review]
LGTM. Thanks
Please, wait also for @bpintea's approval before merging.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/CombineBinaryComparisons.java
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/CombineBinaryComparisons.java:49]: I know this is a c&p issue, but reading just the comment hints that Ranges need special handling here, when in fact they're just ignored in theory, but in practice they don't occur at all. I don't think it's worth another CI round, but maybe we can either add a TODO or remove the comment.

**Review by bpintea — COMMENTED** [review]
Why discard the `Range` formation? We're not having language support for smth. like `BETWEEN`, but I think `Range`s could very well be used internally.

Right now an ESQL query like `from employees | where salary > 30000 and salary  < 50000`
is translated into a DSL `bool` query with two distinctive range queries:
```
query[{"bool":{"must":[{"esql_single_value":{"field":"salary","next":{"range"
:{"salary":{"gt":30000,"boost":1.0}}},"source":"salary > 30000@5:38"}},{"esql_single_value":{"field":"salary","next":{"range":{"salary":{"lt":50000,"boost":1.0}}},"source":"salary  < 50000@5:57"}}],"boost":1.
0}}
```
We can optimise the filter output already and potentially enable other Range-based optimisation rules.

**Comment by fang-xing-esql** [comment]
> Why discard the `Range` formation? We're not having language support for smth. like `BETWEEN`, but I think `Range`s could very well be used internally.
> 
> Right now an ESQL query like `from employees | where salary > 30000 and salary < 50000` is translated into a DSL `bool` query with two distinctive range queries:
> 
> ```
> query[{"bool":{"must":[{"esql_single_value":{"field":"salary","next":{"range"
> :{"salary":{"gt":30000,"boost":1.0}}},"source":"salary > 30000@5:38"}},{"esql_single_value":{"field":"salary","next":{"range":{"salary":{"lt":50000,"boost":1.0}}},"source":"salary  < 50000@5:57"}}],"boost":1.
> 0}}
> ```
> 
> We can optimise the filter output already and potentially enable other Range-based optimisation rules.

That's a good point! Although BETWEEN is not supported in ES|QL yet, Range can be useful for the pushing down predicates in physical plan, I'll look more into it.

**Comment by fang-xing-esql** [comment]
Thank you for reviewing @bpintea! As suggested, I created a separate PR #111255 to add both `CombineBinaryComparisons` and `Range`, it is a more complete solution, more complicated also :). I don't have a strong opinion to either including or not including `Range`, if we are going to support BETWEEN in ES|QL in the future, we will need `Range` at some point.

**Review by bpintea — APPROVED** [review]

**Comment by fang-xing-esql** [comment]
Thank you for reviewing @bpintea @astefan ! It was a long discussion, but hopefully it worths it :-). I'll merge this once CI is clear.

**Pull request merged**
